### ISSUE TITLE
Promoting downstream from control-flow stage doesn't work

### ISSUE THREAD

**Issue by jessesuen** [akuity/kargo#2496]
# Description

When clicking the truck icon on a control-flow stage (one that doesn't have any promotion mechanisms), the UI doesn't highlight any Freight as promotable, so I am unable to promote to all the downstream stages with one click. I'm not sure if this is a UI problem or a backend problem.

The following video shows "Promote downstream" working for `staging`, but not for `prod`.

https://github.com/user-attachments/assets/4efacd6d-a763-4b9f-8482-35e1ce52f15d

Here is the live object of the `prod` Stage:

```yaml
metadata:
  name: prod
  namespace: kargo-advanced
spec:
  subscriptions: {}
  requestedFreight:
    - origin:
        kind: Warehouse
        name: guestbook
      sources:
        direct: false
        stages:
          - ab-test-a
          - ab-test-b
status:
  phase: NotApplicable
  observedGeneration: "2"
  freightSummary: 0/1 Fulfilled
```

# Steps to Reproduce

Create the [kargo-advanced](https://github.com/akuity/kargo-advanced) example and try to promote downstream from prod stage.

# Version

```
v0.8.7
```

**Comment by Marvin9** [comment]
I was able to reproduce. Then tried to check if this is UI bug, I don't know the reason but UI logic says if stage has only 1 downstream subscriber then don't let promote. I changed that to let it promote but then saw something in downstream promotion API

https://github.com/akuity/kargo/blob/main/internal/api/promote_downstream_v1alpha1.go#L137-L139

```
// Avoid creating a Promotion if the downstream Stage has no
// PromotionMechanisms, and is a "control flow" Stage.
```

In your example @jessesuen , `prod` stage has no promotion mechanism and for that reason even if we have UI fixed, API won't let promote. 

@krancour can you please chime in.

**Comment by krancour** [comment]
I will look at this today.

**Comment by krancour** [comment]
> I don't know the reason but UI logic says if stage has only 1 downstream subscriber then don't let promote.

It's because of https://github.com/akuity/kargo/issues/2247

Truck deliberately only appears if there are multiple downstream Stages.

I'm not sure that's related to the issue at hand.

So far I cannot find anything in the back end that would explain this.

@jessesuen are there any any errors in the JS console when this happens?

**Comment by Marvin9** [comment]
prod can't send freight to downstream because it doesn't have any freight available.

prod can't have freight because ab-test won't downstream for 2 reasons

1. UI interpretation - No downstream in 1:1, only allows for 1:N if N > 1. Event if UI lets downstream, the operation won't happen because the API code I linked would disallow downstream promotion if stage doesn't have promotionMechanism which is the case for prod.
2. UI interpretation - No upstream freight for prod because it doesn't have promotionMechanism (Now I don't know the reason for this interpretation but this could be the UI bug if the interpretation is wrong)

We could reach to the solution if my theory for 2nd reason is correct.

It looks like to me that `control-flow` is just concept and at some point with some breaking changes, UI lost the definition of what control-flow is. @krancour you are the best person to enlighten on this.

**Comment by krancour** [comment]
> It looks like to me that control-flow is just concept and at some point with some breaking changes, UI lost the definition of what control-flow is.

Historically, a "control flow stage" isn't really anything except a Stage that lacks any promotion mechanisms.

For v0.9.0, this definition needs to change to "doesn't have any (legacy/deprecated) promotion mechanisms _and_ doesn't have a promotion template."

It's quite probable that the UI is not checking for the second condition in a few places. I've occasionally been noticing places in the back-end where we are also not accounting for it.

In v1.0.0, the deprecated promotion mechanisms will be fully removed and the definition of a control flow stage will be just "doesn't have a promotion template."

All of this being said, I don't think @jessesuen's issue is related to this (or at least not _directly_), as he opened this issue prior to the support for promotion template / steps being merged.

**Comment by krancour** [comment]
Ok... this was a UI bug introduced way back in #2298.

That PR deprecated `stage.status.currentFreight` and in the same PR, @rbreeze created a `getCurrentFreight()` func in the UI that gets the current Freight from the top of the `stage.status.freightHistory` stack.

It seems this func or calls made on its results are not adequately handling control flow stages that never have _any_ freight.

**Comment by Marvin9** [comment]
This was fixed by https://github.com/akuity/kargo/pull/2497 and is available in release greater than 0.8.7 but I am still checking for anomalies will close this if everything looks correct

**Comment by Marvin9** [comment]
Closing this issue

https://github.com/user-attachments/assets/dd6f98eb-3c0a-4b31-b521-c465ea0f596a
